kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,437 @@
+    val buffer = collection.mutable.Buffer[FilteredColumnarBatch]()
+    while (paginatedIter.hasNext) {
+      val batch = paginatedIter.next()
+      buffer += batch
+    }
+    buffer
+  }
+
+  private def validatePageResults(
+      batches: Seq[FilteredColumnarBatch],
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      testName: String): Unit = {
+    assert(batches.nonEmpty)
+    val fileCounts: Seq[Long] = batches.map(_.getPreComputedNumSelectedRows.get().toLong)
+    val totalFileCountsReturned = fileCounts.sum
+
+    assert(fileCounts.length == expectedBatchCount)
+    assert(totalFileCountsReturned == expectedFileCount)
+
+    logger.info(s"$testName: Total num batches returned in page one = ${fileCounts.length}")
+    logger.info(s"$testName: Total num Parquet Files fetched in page one = " +
+      s"$totalFileCountsReturned")
+  }
+
+  private def validatePageToken(
+      paginatedIter: PaginatedScanFilesIterator,
+      expectedLogFilePath: String,
+      expectedRowIndex: Long,
+      testName: String): Unit = {
+    val lastReadLogFilePath = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReadLogFilePath
+    val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReturnedRowIndex
+
+    assert(lastReadLogFilePath.endsWith(expectedLogFilePath))
+    assert(lastReturnedRowIndex == expectedRowIndex)
+
+    logger.info(s"$testName: New PageToken: lastReadLogFileName = $lastReadLogFilePath")
+    logger.info(s"$testName: New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+  }
+
+  private def runSinglePaginationTestCase(
+      testName: String,
+      tablePath: String,
+      tableVersionOpt: Optional[Long],
+      pageSize: Long,
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      expectedLogFileName: String,
+      expectedRowIndex: Int): Unit = {
+    val paginatedScan = createPaginatedScan(
+      tablePath = tablePath,
+      tableVersionOpt = tableVersionOpt,
+      pageSize = pageSize)
+
+    val paginatedIter = paginatedScan.getScanFiles(customEngine)
+    val returnedBatchesInPage = collectPaginatedBatches(paginatedIter)
+
+    validatePageResults(returnedBatchesInPage, expectedFileCount, expectedBatchCount, testName)
+    validatePageToken(paginatedIter, expectedLogFileName, expectedRowIndex, testName)
+
+    paginatedIter.close()
+  }
+
+  // TODO: test call hasNext() twice
+
+  // File name constants for better readability
+  private val JSON_FILE_0 = "00000000000000000000.json"
+  private val JSON_FILE_1 = "00000000000000000001.json"
+  private val JSON_FILE_2 = "00000000000000000002.json"
+  private val JSON_FILE_11 = "00000000000000000011.json"
+  private val JSON_FILE_12 = "00000000000000000012.json"
+  private val CHECKPOINT_FILE_10 = "00000000000000000010.checkpoint.parquet"
+
+  // Single JSON file test cases
+  val singleJsonTestCases = Seq(
```

Comment:
oh, you can actually just delete this `val singleJsonTestCases` and just declare the `Seq` list and then do the `.foreach { testCase => ...`

and then yeah, keep the comment for what this is for. `// ===== Single JSON file test cases =====`